crates/server/src/server/h3_handler.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2015-2021 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{io, net::SocketAddr, sync::Arc};
+
+use bytes::{Buf, Bytes};
+use drain::Watch;
+use futures_util::lock::Mutex;
+use h3::server::RequestStream;
+use h3_quinn::BidiStream;
+use tracing::{debug, warn};
+use trust_dns_proto::{
+    error::ProtoError, h3::h3_server::H3Connection, http::Version, https::HttpsError, rr::Record,
+};
+
+use crate::{
+    authority::MessageResponse,
+    server::{
+        request_handler::RequestHandler, response_handler::ResponseHandler, server_future,
+        Protocol, ResponseInfo,
+    },
+};
+
+pub(crate) async fn h3_handler<T>(
+    handler: Arc<T>,
+    mut connection: H3Connection,
+    src_addr: SocketAddr,
+    _dns_hostname: Option<Arc<str>>,
+    shutdown: Watch,
+) -> Result<(), ProtoError>
+where
+    T: RequestHandler,
+{
+    // TODO: we should make this configurable
+    let mut max_requests = 100u32;
+
+    // Accept all inbound requests sent over the connection.
+    loop {
+        let (_, mut stream) = tokio::select! {
+            result = connection.accept() => match result {
+                Some(Ok(next_request)) => next_request,
+                Some(Err(err)) => {
+                    warn!("error accepting request {}: {}", src_addr, err);
+                    return Err(err);
+                }
+                None => {
+                    break;
+                }
+            },
+            _ = shutdown.clone().signaled() => {
+                // A graceful shutdown was initiated.
+                break;
+            },
+        };
+
+        let request = match stream
+            .recv_data()
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 stream receive data failed: {e}")))?
+        {
+            Some(mut request) => request.copy_to_bytes(request.remaining()),
+            None => {
+                warn!("h3 stream received no data");
+                continue;
+            }
+        };
+
+        debug!(
+            "Received bytes {} from {src_addr} {request:?}",
+            request.remaining()
+        );
+        let handler = handler.clone();
+        let stream = Arc::new(Mutex::new(stream));
+        let responder = H3ResponseHandle(stream.clone());
+
+        handle_request(request, src_addr, handler, responder).await;
+
+        max_requests -= 1;
+        if max_requests == 0 {
+            warn!("exceeded request count, shutting down h3 conn: {src_addr}");
+            connection.shutdown().await?;
```

Comment:
Unlike DoQ I'm shutting down the whole connection here instead of the stream.

Reply:
Any particular reason? I'm ok with that, but I would like to understand. I think in h2 we spawn on each new stream so we can handle parallel requests from a single source, seems like it might be easy enough to do? see: https://github.com/bluejekyll/trust-dns/blob/main/crates/server/src/server/https_handler.rs#L73-L78

Reply:
> Any particular reason?

No, I just didn't understand why DoQ only closes the stream, because it exits the loop anyway, so new streams aren't handled anymore.

> I think in h2 we spawn on each new stream so we can handle parallel requests from a single source, seems like it might be easy enough to do?

I can do that as well, but DoH doesn't have the request limit, so should I remove that limit or do both? Presumably with an `AtomicUsize`.

Reply:
I'm ok with us leaving this as is, and just leaving a note that there is a potential optimization here for concurrent requests on the same connection. Something we can do in a followup PR.

Reply:
I don't know why it seems so complicated in my head, but I just wrapped the `handle_request()` function in `tokio::spawn()` and left the rest as-is.